src/Symfony/Bridge/Doctrine/Form/ChoiceList/EntityChoiceList.php

```diff
@@ -107,16 +114,21 @@ public function __construct(ObjectManager $manager, $class, $labelPath = null, E
         $this->class = $this->classMetadata->getName();
         $this->loaded = is_array($entities) || $entities instanceof \Traversable;
         $this->preferredEntities = $preferredEntities;
+        list(
+            $this->idField,
+            $this->idAsIndex,
+            $this->idAsValue
+        ) = $this->getIdentifierInfoForClass($this->classMetadata);
+
+        if (null !== $this->idField && $this->classMetadata->hasAssociation($this->idField)) {
+            $this->idClassMetadata = $this->em->getClassMetadata(
+                $this->classMetadata->getAssociationTargetClass($this->idField)
+            );
 
-        $identifier = $this->classMetadata->getIdentifierFieldNames();
-
-        if (1 === count($identifier)) {
-            $this->idField = $identifier[0];
-            $this->idAsValue = true;
-
-            if (in_array($this->classMetadata->getTypeOfField($this->idField), array('integer', 'smallint', 'bigint'))) {
-                $this->idAsIndex = true;
-            }
+            list(
+                $this->idAsIndex,
+                $this->idAsValue
+            ) = array_slice($this->getIdentifierInfoForClass($this->idClassMetadata), 1);
```

Comment:
The `array_slice` call is required (or a similar approach obviously :), since when called the second time the returned identifier is the one from the associated entity. The field name may be different in the two entities, so it is important that only the type information are carried on, without the foreign name.

I can refactor the method as `initIdentifierInfoForClass`, but still I have to do something like:
```
$bakIdField = $this->idField;
$this->initIdentifierInfoForClass($this->idClassMetadata); 
$this->idField = $bakIdField;
```
or like:
```
$this->initIdentifierInfoForClass($this->idClassMetadata, false); 
// ...
private function getIdentifierInfoForClass(ClassMetadata $classMetadata, $initIdField)
// ...
```
Which one is clearer? I vote for the second one...